# Описание проекта
Заказчик этого исследования — большая телекоммуникационная компания, которая оказывает услуги на территории всего СНГ. Перед компанией стоит задача определить текущий уровень потребительской лояльности, или NPS (от англ. Net Promoter Score), среди клиентов из России. 

Чтобы определить уровень лояльности, клиентам задавали классический вопрос: «Оцените по шкале от 1 до 10 вероятность того, что вы порекомендуете компанию друзьям и знакомым».

Компания провела опрос и попросила вас подготовить дашборд с его итогами. Большую базу данных для такой задачи разворачивать не стали и выгрузили данные в SQLite. 

Чтобы оценить результаты опроса, оценки обычно делят на три группы:
- 9-10 баллов — «cторонники» (англ. promoters);
- 7-8 баллов — «нейтралы» (англ. passives);
- 0-6 баллов — «критики» (англ. detractors).

Итоговое значение NPS рассчитывается по формуле: % «сторонников» - % «критиков».

Таким образом, значение этого показателя варьируется от -100% (когда все клиенты «критики») до 100% (когда все клиенты лояльны к сервису). Но это крайние случаи, которые редко встретишь на практике. 

Интерпретируя результаты NPS-опросов, следует также помнить, что само значение мало о чём говорит. Однако исследования показывают, что клиенты-сторонники полезны любому бизнесу. Они чаще других повторно совершают покупки, активнее тестируют обновления и приводят в сервис своих друзей и знакомых. Поэтому NPS остаётся одной из важнейших метрик бизнеса. 

# Описание данных

1) Таблица user

Содержит основную информацию о клиентах.

- user_id - Идентификатор клиента, первичный ключ таблицы
- lt_day -	Количество дней «жизни» клиента
- age - Возраст клиента в годах
- gender_segment -	Пол клиента (1 – женщина, 0 – мужчина)
- os_name -	Тип операционной системы
- cpe_type_name - 	Тип устройства
- location_id	- Идентификатор домашнего региона клиента, внешний ключ, отсылающий к таблице location
- age_gr_id - 	Идентификатор возрастного сегмента клиента, внешний ключ, отсылающий к таблице age_segment
- tr_gr_id - Идентификатор сегмента клиента по объёму потребляемого трафика в месяц, внешний ключ, отсылающий к таблице traffic_segment
- lt_gr_id - Идентификатор сегмента клиента по количеству месяцев «жизни», внешний ключ, отсылающий к таблице lifetime_segment
- nps_score -	Оценка клиента в NPS-опросе (от 1 до 10)

2) Таблица location

Справочник территорий, в которых телеком-компания оказывает услуги.

- location_id -	Идентификатор записи, первичный ключ
- country -	Страна
- city - Город

3) Таблица age_segment

Данные о возрастных сегментах клиентов.

- age_gr_id	- Идентификатор сегмента, первичный ключ
- bucket_min  - Минимальная граница сегмента
- bucket_max - Максимальная граница сегмента
- title	- Название сегмента

4) Таблица traffic_segment

Данные о выделяемых сегментах по объёму потребляемого трафика.

- tr_gr_id - Идентификатор сегмента, первичный ключ
- bucket_min -	Минимальная граница сегмента
- bucket_max - Максимальная граница сегмента
- title - Название сегмента

5) Таблица lifetime_segment

Данные о выделяемых сегментах по количеству месяцев «жизни» клиента — лайфтайму.

-  lt_gr_id - Идентификатор сегмента, первичный ключ
- bucket_min -	Минимальная граница сегмента
- bucket_max -	Максимальная граница сегмента
- title -	Название сегмента


# План исследования

**Шаг 1.** Подключение к базе

Прежде чем начать работу, нам необходимо получить доступ к базе данных. Данные выгрузили в SQLite — СУБД, в которой база данных представлена файлом. Для подключения к такой базе достаточно иметь доступ к файлу с расширением .db.

Путь к файлу: /datasets/telecomm_csi.db. 

**Шаг 2.** Выгрузка данных

- Теперь нужно подготовить данные для дашборда. Преобразовывать данные с помощью Python нельзя — мы должны использовать только SQL-запросы. 
- Нам нужно собрать в одну витрину данные из разных таблиц. Эту витрину мы будем использовать для дашборда.

- Определить, является ли клиент новым, можно по количеству дней «жизни». Такой показатель ещё называют лайфтайм. Его отсчитывают с момента совершения важного события для бизнеса, например покупки сим-карты для телеком-компании. Клиента считают новым, если количество дней «жизни» составляет не более 365 дней. 
- Получившуюся таблицу сохраните как CSV-файл.

**Шаг 3.** Создание дашборда в Tableau

- Нам нужно построить дашборд, который представит информацию о текущем уровне NPS среди клиентов и покажет, как этот уровень меняется в зависимости от пользовательских признаков. 
- Из дашборда должно быть понятно, какие группы пользователей участвовали в опросе. Обратим внимание, что если пользователи какой-то группы преобладают, то это могло отразиться на общем уровне NPS. 

**Шаг 4.** Ответы на вопросы с помощью дашборда

Используя дашборд, ответим на вопросы:
- Как распределены участники опроса по возрасту и полу? Каких пользователей больше: новых или старых? Пользователи из каких городов активнее участвовали в опросе?
- Какие группы пользователей наиболее лояльны к сервису? Какие менее?
- Какой общий NPS среди всех опрошенных?
- Как можно описать клиентов, которые относятся к группе cторонников (англ. promoters)?

Ответы на вопросы оформим в виде презентации. 


In [1]:
# импортируем нужные библиотеки
import os
import pandas as pd
import numpy as np

from sqlalchemy import create_engine

In [2]:
# путь к БД на нашем компьютере (например, в той же папке, что и тетрадь)
path_to_db_local = 'telecomm_csi.db'
# путь к БД на платформе
path_to_db_platform = '/datasets/telecomm_csi.db'
# итоговый путь к БД
path_to_db = None

# если путь на нашем компьютере ведёт к БД, то он становится итоговым
if os.path.exists(path_to_db_local):
    path_to_db = path_to_db_local
# иначе: если путь на платформе ведёт к БД, то он становится итоговым
elif os.path.exists(path_to_db_platform):
    path_to_db = path_to_db_platform
# иначе выводится сообщение о том, что файл не найден
else:
    raise Exception('Файл с базой данных SQLite не найден!')

# если итоговый путь не пустой
if path_to_db:
    # то создаём подключение к базе
    engine = create_engine(f'sqlite:///{path_to_db}', echo=False)
    

Добавим следующие поля:
- age_segment -	Возрастной сегмент
- traffic_segment -	Сегмент по объёму потребляемого трафика
- lifetime_segment - Сегмент по количеству дней «жизни»
- nps_group - Поле хранит информацию о том, к какой группе относится оценка клиента в опросе


In [8]:
# создаем строковую переменную в Python, которая содержит SQL-запрос.
query = """
SELECT  u.user_id as user_id,
        (CASE
           WHEN u.lt_day > 365 THEN 'старый'
           WHEN u.lt_day <= 365 THEN 'новый'
        END) as is_new,
        u.lt_day as lt_day,
        u.age as age,
        (Case 
            WHEN u.gender_segment = 1 THEN 'женщина'
            WHEN u.gender_segment = 0 THEN 'мужчина'
            ELSE 'Нет информации'
        END) as gender_segment, 
        u.os_name as os_name, 
        u.cpe_type_name as cpe_type_name,
        lt.country as country,
        lt.city as city,
        ags.title as age_segment,
        ts.title as traffic_segment,
        ls.title as lifetime_segment,
        u.nps_score as nps_score,
        (CASE
           WHEN u.nps_score BETWEEN 9 AND 10 THEN 'cторонники'
           WHEN u.nps_score BETWEEN 7 AND 8 THEN 'нейтралы'
           WHEN u.nps_score BETWEEN 0 AND 6  THEN 'критики'
        END) as nps_group
FROM user as u
LEFT JOIN location as lt on u.location_id = lt.location_id
LEFT JOIN age_segment as ags on u.age_gr_id = ags.age_gr_id
LEFT JOIN traffic_segment as ts on u.tr_gr_id = ts.tr_gr_id
LEFT JOIN lifetime_segment as ls on u.lt_gr_id = ls.lt_gr_id;

"""

In [4]:
# создаём датафрейм по данным запроса
df = pd.read_sql(query, engine)
df.head(3)

,user_id,is_new,lt_day,age,gender_segment,os_name,cpe_type_name,country,city,age_segment,traffic_segment,lifetime_segment,nps_score,nps_group
0,A001A2,старый,2320,45.0,женщина,ANDROID,SMARTPHONE,Россия,Уфа,05 45-54,04 1-5,08 36+,10,cторонники
1,A001WF,старый,2344,53.0,мужчина,ANDROID,SMARTPHONE,Россия,Киров,05 45-54,04 1-5,08 36+,10,cторонники
2,A003Q7,старый,467,57.0,мужчина,ANDROID,SMARTPHONE,Россия,Москва,06 55-64,08 20-25,06 13-24,10,cторонники


In [5]:
#выводим общую информацию
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 502493 entries, 0 to 502492
Data columns (total 14 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   user_id           502493 non-null  object 
 1   is_new            502493 non-null  object 
 2   lt_day            502493 non-null  int64  
 3   age               501939 non-null  float64
 4   gender_segment    502493 non-null  object 
 5   os_name           502493 non-null  object 
 6   cpe_type_name     502493 non-null  object 
 7   country           502493 non-null  object 
 8   city              502493 non-null  object 
 9   age_segment       502493 non-null  object 
 10  traffic_segment   502493 non-null  object 
 11  lifetime_segment  502493 non-null  object 
 12  nps_score         502493 non-null  int64  
 13  nps_group         502493 non-null  object 
dtypes: float64(1), int64(2), object(11)
memory usage: 53.7+ MB


In [6]:
# выводим статистическое описание данных
df.describe()

,lt_day,age,nps_score
count,502493.000000,501939.000000,502493.000000
mean,1868.841439,39.621946,7.508562
std,1683.701762,11.188249,3.020378
min,-21.000000,10.000000,1.000000
25%,533.000000,31.000000,5.000000
50%,1239.000000,38.000000,9.000000
75%,3064.000000,47.000000,10.000000
max,9162.000000,89.000000,10.000000


In [7]:
# сохраненяем df в CSV-файл без индексов
df.to_csv('telecomm_csi_tableau.csv', index=False)

Далее создаем дашборд в Tableau на основе подготовленного датасета и используем его для анализа данных и ответов на исследовательские вопросы.

Ссылка на презентацию :https://public.tableau.com/shared/BSQB958GX?:display_count=n&:origin=viz_share_link

Общие выводы и наблюдения: 
- По полу преобладают женщины(~54% от общего числа опрошенных)

- По возрасту: большинство клиентов в возрасте от 25 до 45 лет

- Старых клиентов почти в 5 раз больше чем новых(415к против 85к)

- Наиболее активные участники опроса -  жители Москвы(~ 65к), Санкт-Петербурга(~ 29к) и Новосибирска(~26к)

- Сторонниками сервиса в основном являются женщины возрастной группы от 25 до 45 лет и большинство - старые пользователи. Но и среди критиков в основном та же возрастная группа и также старые пользователи. Это может быть связано с тем, что кол-во старых пользователей в 5 раз больше, чем новых, поэтому их доля доминирует как среди лояльных, так и среди нелояльных клиентов. Следовательно, выборка неравномерна по стажу, и это нужно учитывать при интерпретации выводов о лояльности.

- Общий NPS среди всех опрошенных равен 21,97%

- По NPS наиболее лояльные новые пользователи: мужчины в возрасте от 45 до 64 и женщины от 25 до 34 и от 45 до 54. По устройству пользователи ANDROID более лояльны, чем пользователи IOS

- Сторонников больше среди старых пользователей в абсолютных числах — потому что их больше в выборке. Но относительно (в процентах) — новые пользователи лояльнее. Это и показывает более высокий NPS у новых.



Таким образом, можно сделать вывод, что основную аудиторию сервиса составляют женщины в возрасте от 25 до 45 лет, преимущественно старые пользователи. Несмотря на то, что сторонников в абсолютных числах больше среди старых клиентов, именно новые пользователи демонстрируют наибольшую лояльность — это отражается в более высоком NPS.

Также выявлены возрастно-половые и поведенческие сегменты с наибольшей лояльностью — это может быть использовано для настройки маркетинговых коммуникаций и удержания. При дальнейшем анализе рекомендуется учитывать неравномерность выборки по стажу, чтобы получать более точные и репрезентативные оценки клиентского отношения.